inline.h

```diff
@@ -2618,23 +2618,30 @@ Perl_mortal_getenv(const char * str)
 {
     /* This implements a (mostly) thread-safe, sequential-call-safe getenv().
      *
-     * It's (mostly) thread-safe because it uses a mutex to prevent
-     * simultaneous access from other threads that use the same mutex, and
-     * makes a copy of the result before releasing that mutex.  All of the Perl
-     * core uses that mutex, but, like all mutexes, everything has to cooperate
-     * for it to completely work.  It is possible for code from, say XS, to not
-     * use this mutex, defeating the safety.
+     * It's (mostly) thread-safe because it uses a mutex to prevent other
+     * threads (that look at this mutex) from destroying the result before this
+     * routine has a chance to copy the result to a place that won't be
+     * destroyed before the caller gets a chance to handle it.  That place is a
+     * mortal SV.  khw chose this over SAVEFREEPV because he is under the
+     * impression that the SV will hang around longer under more circumstances
      *
-     * On some platforms, getenv() is not sequential-call-safe, because
-     * subsequent calls destroy the static storage inside the C library
-     * returned by an earlier call.  The result must be copied or completely
-     * acted upon before a subsequent getenv call.  Those calls could come from
-     * another thread.  Again, making a copy while controlling the mutex
-     * prevents these problems..
+     * The reason it isn't completely thread-safe is that other code could
+     * simply not pay attention to the mutex.  All of the Perl core uses the
+     * mutex, but it is possible for code from, say XS, to not use this mutex,
+     * defeating the safety.
+     * 
+     * getenv() returns, in some implementations, a pointer to a spot in the
+     * **environ array, which could be invalidated at any time by this or
+     * another thread changing the environment.  Other implementations copy the
+     * **environ value to a static buffer, returning a pointer to that.  That
+     * may be a per-process rather than per-thread buffer, in which case this
+     * routine could be destroying another thread's value, and we need an
+     * exclusive lock.  Otherwise, a many-reader lock is ok.  There is a
```

Comment:
This claims there might be a single buffer, but that isn't what POSIX says:

> The returned string pointer might be invalidated or [Option End]  the string content might be overwritten by a subsequent call to getenv(), [CX] [Option Start] setenv(), unsetenv(), [Option End]
[XSI] [Option Start]  or (if supported) putenv() [Option End]  but they shall not be affected by a call to any other function in this volume of POSIX.1-2017.

It might be a static buffer, it might be a malloced pointer which is then freed on the next call to getenv().

It doesn't (I think) make any difference for this implementation, might it might mislead someone into thinking the pointer from getenv() remains valid but pointing to a changed buffer, when this isn't the case.

Reply:
I can generalize it.

It doesn't make a difference to this implementation, provided the probe worked correctly, and that every thread is using this mutex.  Mutexes only work on Posix boxes as long as everyone agrees to use them, and doesn't try an unauthorized access.  So, if another thread calls getenv directly without the mutex, it could be a problem if that getenv destroyed this one.  But we have that problem in existing code, and this mutex should remove it as a problem for all cooperating code.

Reply:
The new version has changed the wording